### ISSUE TITLE
TTV's violate certain laws of physics

### ISSUE THREAD

**Issue by Robustin** [tgstation/tgstation#15534]
3rd attempt to post this issue:

You can close it once you answer these three questions in a way that explains why they aren't bugs:

1) Mixing oxygen and hot plasma/co2 at low pressure values with a TTV results in the plasma and co2 completely disappearing from the mix and no explosion resulting. You get two tanks of warm oxygen and a flagrant violation of the laws of physics.

Bonus Proof Screenshot because nobody reads shit unless its in an image: 

![sdy8mzb](https://cloud.githubusercontent.com/assets/7305553/13036906/881eedb8-d36b-11e5-85ec-7c2c0087d661.png)

2) Mixing ultrahot pure plasma and chilled oxygen blows up a single tank of the TTV with no actual explosion/sound/effect and the sprite still looks like both tanks are attached.

3) Other assorted shit like how canister combustion won't occur when its still wrench to a valve but as soon as you unwrench it it burns.

**Comment by WJohn** [comment]
Toxins: perpetually broken.

Just like turrets. And often R&D linking.

**Comment by PKPenguin321** [comment]
this one's still open fam: https://github.com/tgstation/-tg-station/issues/15531

**Comment by duncathan** [comment]
It's literally still open dude WTF
Anyway I'll close the old one cuz this one has a picture

**Comment by as334** [comment]
Can we get a BLAME MSO tag, since the bombs nerf thing was him.

**Comment by phil235** [comment]
The plasma doesn't completely disappear, there's just very little of it in the first place. Going by the numbers in your screenshots there's 120 times more oxygen than plasma in total so it's normal to end up with less than one percent plasma in your final mix.

**Comment by Dorsisdwarf** [comment]
I replaced your shitty imgur link with an inline version, ooh, look that would have been _easier_ than using imgur

**Comment by octareenroon91** [comment]
OK, initial conditions:
Plasma tank : 0.097 moles of hot gas 
Oxygen tank : 11.510 moles of cold gas

Final state:
5.799 moles in each of the two tanks = 11.598 moles combined.

Mass is conserved to within 1 part in 1000 . The "Oxygen 99%" on the final tanks is evidence enough that there's "something else" in the tanks.  The other two are zero due to round-off before display.

**Comment by Robustin** [comment]
Alright I get the first one Phil/Octa, didn't think that the temp would have that dramatic of an effect. Thanks for doing the math. 

I was only getting frustrated because the first two closures were basically "lol u cant get maxcap". When something like "the hot gas is a tiny amount compared to the chilled oxygen" would've addressed the first issue much more gracefully. 

Regarding number 2, I'm aware of tanks popping at a certain pressure, but these weren't those kind of situations. When I manage to make a tank pop its something that happens within seconds, these incidents only happened when the valves opened and prior to the test the tank was clocking like 1200-1300kpa and slowly climbing. At the very least the sprite issue showing both tanks intact is a bug.

There's still a whole lot of unexplained shit. I can't figure out why putting 200 degree plasma into a regular oxygen tank gives you a grenade with about 8 seconds until it punches a hole in the hull, but heating a 50/50 oxygen/plasma mix to 200 degrees and then filling an empty tank with it gives you a very, very slow combustion that takes like one minute to hit max pressure and even then it just leaks for a bit until the plasma finishes burning.

**Comment by Robustin** [comment]
Just tried another run with a superchilled oxygen tank (used an upgraded fridge and a pump so each bomb literally required 1.5 canisters of oxygen) and an ideally heated plasma tank.

Tried to set it off, nothing. Got the tanks checked, both are fucking empty.

**Comment by duncathan** [comment]
What's ideally heated? How do you, physically or code-wise, know what heat is ideal?

Prior experience doesn't cut it; unless you're Cuban Pete, the bombcap was always too low to encourage actually decent mixes. The wiki had been propagating a bad mix for as long as I've been aware.

**Comment by Actionb** [comment]
Feel free to chime in and call me an idiot if all this is incorrect. Most of this is from a physical point of view and has only little to do with in-game stuff:

Keep in mind that having a good amount of reactants is more important than having extreme temperature differences. What makes an explosion is a sudden, runaway change in energy; and when you have just a very few moles of superhot plasma to react with a LOT of moles of very cool oxygen, you won't get a whole lot of energy before all the plasma has been consumed. The remaining oxygen just makes it worse by cooling the endproduct and slowing reaction speed.

Ideally you would want the resulting mixture to be 1:1 moles of plasma:oxygen with a temperature above the ignition point. This would allow all moles of plasma to find a partner to react with and to release their energy in the conversion to CO2. Make the plasma as hot and the oxygen as cool as possible while still maintaining the balance for extra sudden energy releases (by having the plasma suddenly cool down upon contact).
Obviously, having the initial mixture start at 101°C won't make for an impressive display; the more energy (in this case in form of temperature) is availble to the surrounding reactants, the faster the reaction... this is the thermal runaway part.

Afaik, what makes the explosion in-game is actually the tank bursting from pressure - independant from what happens inside the tank. So you can deviate from the 'ideal' mixture all you want, as long as the reaction is enough to cause a violent increase in pressure.
And considering this is a game and not the real world, there might be more limitations that I am not aware of (like tanks self-destructing without causing an explosion).

Bottom line is: getting a good explosion isn't about having the most extreme numbers, but more about having a moderate mixture that leads to the result you want.
I guess this isn't immediately obvious to most toxins scientists, leading to confusion.

> (used an upgraded fridge and a pump so each bomb literally required 1.5 canisters of oxygen)
Tried to set it off, nothing. Got the tanks checked, both are fucking empty.

My assumption: Mixing the plasma with all that oxygen resulted in a minor reaction that wasn't enough to cause a detonation, but was enough to raise pressure due to expanding oxygen in the tanks to cause them to leak.

**Comment by duncathan** [comment]
That's 100% correct, Actionb.

**Comment by Saegrimr** [comment]
I have a related question.

http://i.imgur.com/dYq4CUJ.png

**Comment by Robustin** [comment]
Ideally heated as in, hot enough to combust without dilluting the gases through excessive heat or co2. Either way having BOTH tanks completely empty out is not what should ever happen with TTV's once the valve opens.

New Issue: Detonated a large-cap bomb after a rev round had just ended. Kill the captain next to me instantly, but me, the guy with the bomb in my bag, survived with perfect health.

**Comment by AuroranAI** [comment]
Actually, having one tank completely empty out is exactly what should happen. Code wise, think of TTV as a pump that injects the content of the second tank one rapidly into the first tank. To simulate it being an open valve, it magically adds the volume of the second tank to the first tank. This is why when the tank pops, only one of the tanks actually bursts. The TTV not updating when it pops is still a bug, though.

**Comment by Rinoka** [comment]
This is still open, so:

1. Already handled.

2. Pretty much already handled. It's worth noting that superchilling the oxygen in this case means you have a lot of moles, so what you might actually be seeing is not enough plasma to continue the reaction, then temps average out, and the tank explodes because of pressure of the averaged oxygen that remained.

3. I have been skimming and I *believe* this is due to how atmos code treats canisters. Plasma burning initially assigns itself a plasma_burn_rate of 0, and after that it conditionally changes temperature_scale to 1 -- and if the check is invalid (temperature_scale is less than or equal to 0), then it won't let the rest of the simulation run, at all.

temperature_scale defaults (if the above fails) to [temperature - PLASMA_MINIMUM_BURN_TEMPERATURE] / [PLASMA_UPPER_TEMPERATURE - PLASMA_MINIMUM_BURN_TEMPERATURE]. Simplified, that's (temperature - 100+T0C)/(1370+T0C - 100+T0C).

Or, in short, you could effectively compare it to temperature / 1370+T0C.  I will admit I don't know that much here, but I can only come up with two reasons for the no-sell we get with pipenets and canisters.

1) Canisters never call react() unless connected to a preexisting pipenet.

2) Canisters do something wonky to the temperature of their contents, and this error is greatly diminished by being connected to a pipenet.

Honestly, I suspect the former for clear performance reasons, but not quite sure.

**Comment by kriskog** [comment]
^What he said.

3 is pretty much a feature now, but either way I see no benefit from tracking it in this issue.
